### DISCUSSION TITLE
Iteration produces unexpected number of elements

### DISCUSSION THREAD

**Discussion by fardin01** [open-policy-agent/community#439] [OPA and Rego] [answered]
I am writing a policy that looks for given annotations on a K8s object and compares them to a list of required annotations to see if a required annotation is missing. The code can be found [here](https://play.openpolicyagent.org/p/sq2hIyMF8s).

I've tried two different ways:
```
missing := required - provided	
count(missing) > 0
```
and
```
r := required[_]
missing := r; not r in provided
```
They both have the desired outcome. However, in the first case, `count(missing)` returns `1` which is the expected number because there is 1 required annotation and it is missing. In the second case, `count(missing)` returns `10`. I do understand that in this case, I am doing a loop but I do not understand where the number `10` comes from. 

What's happening here?

**Comment by srenatus** [comment, marked as answer]
I believe the misunderstanding might be in the meaning of `;` in 
```rego
r := required[_]
missing := r; not r in provided
```

It does **not** constitute a restriction or WHERE-clause, or the RHS of a comprehension. It **merely** is the same as a line-break. So if you look at it like that
```rego
r := required[_]
missing := r
not r in provided
```
it would make sense that `missing` has the same number of elements as _some element of `required`_. Try adding a `print(missing)` and you'll find that `missing` is a string, `"repository`, for which `count()` returns the number of characters, 10.

I believe you might be looking for a set comprehension here, like
```rego
missing := { r | r := required[_]; not r in provided }
```
or, equivalently,
```rego
missing := { r | some r in required; not r in provided }
```
